ts-sdk/whirlpool/src/harvest.ts

```diff
@@ -0,0 +1,216 @@
+  return {
+    tokenA,
+    tokenB,
+    reward1,
+    reward2,
+    reward3,
+  };
+}
+
+export async function harvestPositionInstructions(
+  rpc: Rpc<GetAccountInfoApi & GetMultipleAccountsApi & GetMinimumBalanceForRentExemptionApi & GetEpochInfoApi>,
+  positionMint: Address,
+  authority: TransactionPartialSigner = DEFAULT_FUNDER,
+): Promise<HarvestPositionInstructions> {
+  invariant(
+    authority.address !== DEFAULT_ADDRESS,
+    "Either supply an authority or set the default funder",
+  );
+  const instructions: IInstruction[] = [];
+
+  const positionAddress = await getPositionAddress(positionMint);
+  const position = await fetchPosition(rpc, positionAddress[0]);
+  const whirlpool = await fetchWhirlpool(rpc, position.data.whirlpool);
+
+  const lowerTickArrayStartIndex = getTickArrayStartTickIndex(
+    position.data.tickLowerIndex,
+    whirlpool.data.tickSpacing,
+  );
+  const upperTickArrayStartIndex = getTickArrayStartTickIndex(
+    position.data.tickUpperIndex,
+    whirlpool.data.tickSpacing,
+  );
+
+  const [positionTokenAccount, lowerTickArrayAddress, upperTickArrayAddress] =
+    await Promise.all([
+      findAssociatedTokenPda({
+        owner: authority.address,
+        mint: positionMint,
+        tokenProgram: TOKEN_PROGRAM_ADDRESS,
+      }).then((x) => x[0]),
+      getTickArrayAddress(whirlpool.address, lowerTickArrayStartIndex).then(
+        (x) => x[0],
+      ),
+      getTickArrayAddress(whirlpool.address, upperTickArrayStartIndex).then(
+        (x) => x[0],
+      ),
+    ]);
+
+  const [lowerTickArray, upperTickArray] = await fetchAllTickArray(rpc, [
+    lowerTickArrayAddress,
+    upperTickArrayAddress,
+  ]);
+
+  const lowerTick = lowerTickArray.data.ticks[getTickIndexInArray(position.data.tickLowerIndex, lowerTickArrayStartIndex, whirlpool.data.tickSpacing)];
+  const upperTick = upperTickArray.data.ticks[getTickIndexInArray(position.data.tickUpperIndex, upperTickArrayStartIndex, whirlpool.data.tickSpacing)];
+
+  const transferFees = await getTransferFeeConfigs(rpc, whirlpool.data);
+
+  const feesQuote = collectFeesQuote(
+    whirlpool.data,
+    position.data,
+    lowerTick,
+    upperTick,
+    transferFees.tokenA,
+    transferFees.tokenB,
+  );
+  const currentUnixTimestamp = BigInt(Math.floor(Date.now() / 1000));
+  const rewardsQuote = collectRewardsQuote(
+    whirlpool.data,
+    position.data,
+    lowerTick,
+    upperTick,
+    currentUnixTimestamp,
+    transferFees.reward1,
+    transferFees.reward2,
+    transferFees.reward3,
+  );
+
+  const { createInstructions, cleanupInstructions, tokenAccountAddresses } =
+    await prepareTokenAccountsInstructions(rpc, authority, [
+      whirlpool.data.tokenMintA,
+      whirlpool.data.tokenMintB,
+      whirlpool.data.rewardInfos[0].mint,
+      whirlpool.data.rewardInfos[1].mint,
+      whirlpool.data.rewardInfos[2].mint,
+    ]);
+
+  instructions.push(...createInstructions);
```

Comment:
We need to execute `updateFeesAndRewards` instruction to move expecting fees and rewards to owed fields.